src/main/java/org/scijava/log/AbstractLogService.java

```diff
@@ -107,7 +125,7 @@ public void removeLogListener(final LogListener l) {
 	}
 
 	@Override
-	public void notifyListeners(final int level, final Object msg,
+	public void notifyListeners(final LogLevel level, final Object msg,
```

Comment:
> The Integer enum pattern has a lot of drawbacks.

So, in general I very much agree with you. However, in this case, I think using integers, or maybe even doubles, actually makes more sense, for similar reasons to why we use doubles for priority (see `org.scijava.Priority`).

> What means level == 0 or LogLevel.NONE? (It's possible meaning "unknown log level" is better expressed by a null reference.)

The idea of 0 i.e. `NONE` is that _nothing_ should be logged, of course. That is different than "unknown" or null. Actually, I would not allow nulls here... another reason `int` happens to be handy, since primitives cannot be null.

> What means level == -1? (Better throw an exception if it's an error.)

I agree that negative levels are not defined here, since 0/NONE is the bottommost behavior. So, we should indeed perform bounds checking and throw an exception if negative values are given anywhere in this API.

> What's the maximum log level? (Now it's clearly LogLevel.TRACE.)

Does there need to be an explicit maximum? I don't think so. It is always possible to be "more verbose" than everything else that has ever existed...